crates/revm/revm-inspectors/src/tracing/builder/parity.rs

```diff
@@ -218,13 +231,116 @@ impl ParityTraceBuilder {
+    pub fn vm_trace(&self) -> VmTrace {
+        let mut walker = CallTraceNodeWalker::new(&self.nodes);
+
+        match walker.next() {
+            Some(start) => Self::make_trace(&mut walker, start),
+            // todo::n should probably but some logging here
+            None => VmTrace { code: Default::default(), ops: Vec::new() },
+        }
+    }
+
+    /// returns a VM trace without the code filled in
+    fn make_trace(
+        walker: &mut CallTraceNodeWalker<'_, DFWalk>,
+        current: &CallTraceNode,
+    ) -> VmTrace {
+        let mut instructions: Vec<VmInstruction> = Vec::with_capacity(current.trace.steps.len());
+
+        for step in current.trace.steps.iter() {
+            let maybe_sub = match step.op.u8() {
+                opcode::CALL
+                | opcode::CALLCODE
+                | opcode::DELEGATECALL
+                | opcode::STATICCALL
+                | opcode::CREATE
+                | opcode::CREATE2 => {
+                    let next = walker.next().expect("missing next node");
+                    Some(Self::make_trace(walker, next))
+                }
+                _ => None,
+            };
+
+            instructions.push(Self::make_instruction(step, maybe_sub));
+        }
+
+        VmTrace { code: Default::default(), ops: instructions }
+    }
+
+    /// todo::n config
+    ///
+    /// Creates a VM instruction from a [CallTraceStep] and a [VmTrace] for the subcall if there is one
+    fn make_instruction(step: &CallTraceStep, maybe_sub: Option<VmTrace>) -> VmInstruction {
+        let maybe_storage = match step.storage_change {
+            Some(storage_change) => {
+                Some(StorageDelta { key: storage_change.key, val: storage_change.value })
+            }
+            None => None,
+        };
+
+        let maybe_memory = match step.memory.len() {
+            0 => None,
+            _ => {
+                Some(MemoryDelta { off: step.memory_size, data: step.memory.data().clone().into() })
+            }
+        };
+
+        let maybe_execution = Some(VmExecutedOperation {
+            used: step.gas_cost,
+            push: match step.new_stack {
+                Some(new_stack) => Some(new_stack.into()),
+                None => None,
+            },
+            mem: maybe_memory,
+            store: maybe_storage,
+        });
+
+        VmInstruction {
+            pc: step.pc,
+            cost: 0, // todo::n
+            ex: maybe_execution,
+            sub: maybe_sub,
+        }
+    }
 }
 
-/// Construct the vmtrace for the entire callgraph
-fn vm_trace(nodes: &[CallTraceNode]) -> VmTrace {
-    // TODO: populate vm trace
+/// addresses are presorted via DF walk thru [CallTraceNode]s
+///
+/// use recursion to fill the [VmTrace] code fields
+pub(crate) fn populate_vm_trace_bytecodes<DB>(
+    db: &DB,
+    df_addresses: &mut dyn Iterator<Item = Address>,
```

Comment:
do we need dyn dispatch here or can make it generic.

we don't have too many different impls for this I think?